### PR TITLE
dirha_wsj recipe for espnet2

### PR DESCRIPTION
**Pull request by yuekaizhang**
The PR is related to [#1490 ](https://github.com/espnet/espnet/issues/1490).
Result for Beam_Circular_Array w/o LM:

# RESULTS
## Environments
- date: `Fri Feb  7 09:38:30 EST 2020`
- python version: `3.7.3 (default, Mar 27 2019, 22:11:17)  [GCC 7.3.0]`
- espnet version: `espnet 0.6.0`
- pytorch version: `pytorch 1.0.1.post2`
- Git hash: `83eb81ca220f8189f94a173851934acf8bbba0df`
  - Commit date: `Tue Jan 28 09:34:59 2020 -0500`

### WER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_dirha_real_Beam_Circular_Arraydecode_lm_valid.loss.best_asr_model_valid.loss.best|409|6762|56.7|40.2|3.1|12.6|55.9|97.3|
|decode_dirha_sim_Beam_Circular_Arraydecode_lm_valid.loss.best_asr_model_valid.loss.best|409|6766|60.2|36.9|2.8|10.1|49.9|96.3|

### CER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_dirha_real_Beam_Circular_Arraydecode_lm_valid.loss.best_asr_model_valid.loss.best|409|39842|77.8|15.1|7.1|12.0|34.2|97.3|
|decode_dirha_sim_Beam_Circular_Arraydecode_lm_valid.loss.best_asr_model_valid.loss.best|409|39869|80.0|13.4|6.6|8.7|28.6|96.3|
#### Todos
- [ ] Rerun experiments for all kinds of mics after word LM is good.  
- [ ] Add Specaug.yaml to local/data.sh

### CODE REVIEW

**Comment by sw005320** [comment]
@sas91, please review it.

**Comment by kamo-naoyuki** [comment]
@yuekaizhang Thank you. I'll implement specaug by different way from espnet1 and it might take about 1~2week. Ignore specaug for now.

File: egs2/dirha_wsj/asr1/run.sh [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+mic=Beam_Circular_Array
+
+local_data_opts="--mic "${mic}" "
+
+stage=9
```
Comment (sas91) [inline egs2/dirha_wsj/asr1/run.sh]: make this stage as "1" when giving PR

File: egs2/dirha_wsj/asr1/asr.sh [hunk 2]
```diff
@@ -0,0 +1 @@
+../../TEMPLATE/asr1/asr.sh
```
Comment (sas91) [inline egs2/dirha_wsj/asr1/asr.sh]: Please add the BLSTM performance with wordlm to RESULTS. Meanwhile I will run this to see if you have missed something.

Comment (yuekaizhang) [reply egs2/dirha_wsj/asr1/asr.sh]: I am not sure whether word LM has been implemented, since egs2/Template/asr.sh wrote" log Word LM is not supported yet". If it has been down, I will run the experiments ASAP. @kamo-naoyuki , do you know whether word LM is good now?

File: egs2/dirha_wsj/asr1/run.sh (hunk 1 above)
Comment (yuekaizhang) [reply egs2/dirha_wsj/asr1/run.sh]: Okay, I'll fix this and other small bugs in CI today.

File: egs2/dirha_wsj/asr1/asr.sh (hunk 2 above)
Comment (kamo-naoyuki) [reply egs2/dirha_wsj/asr1/asr.sh]: Sorry, word lm is not supported.
Espnet2 uses v2 decoding module, but it doesn't support wordLM yet.

**Comment by sw005320** [comment]
Please fix the CI errors in https://travis-ci.org/espnet/espnet/jobs/646232901#L2524-L2548

File: egs2/dirha_wsj/asr1/asr.sh (hunk 2 above)
Comment (yuekaizhang) [reply egs2/dirha_wsj/asr1/asr.sh]: @sas91 , I just added one result to README.md, which is not as good as espnet1. Since it didn't use word lm, the results comparison is not fair. I was wondering whether I need to run the other three mics experiments w/o word lm.

File: egs2/dirha_wsj/asr1/setup.sh [hunk 3]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
+# Set bash to 'debug' mode, it will exit on :
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
```
Comment (kamo-naoyuki) [inline egs2/dirha_wsj/asr1/setup.sh]: You don't need to copy this file.

Comment (yuekaizhang) [reply egs2/dirha_wsj/asr1/setup.sh]: Done. It was a mistake. Thx

File: egs2/dirha_wsj/asr1/conf/train.yaml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+# label smoothing
+#lsm-type: unigram
+#lsm-weight: 0.05
+
+# minibatch related
+batch_size: 15
+max_length:
+     - 800  # maxlen_in if input length  > maxlen_in, batchsize is automatically reduced
+     - 150  # maxlen_out if output length > maxlen_out, batchsize is automatically reduced
```
Comment (kamo-naoyuki) [inline egs2/dirha_wsj/asr1/conf/train.yaml]: You can't specify max_length here because the parameter should be changed according to feats_type.

File: egs2/dirha_wsj/asr1/conf/lm.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+optim: sgd
+patience: 3
+max_length: 150 # 40 for word LMs
```
Comment (kamo-naoyuki) [inline egs2/dirha_wsj/asr1/conf/lm.yaml]: I changed naming rule of config files to test in ci. so please follow #1617

File: egs2/dirha_wsj/asr1/local/data.sh [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+if [ ${stage} -le 1 ] && [ ${stop_stage} -ge 1 ]; then
+    log "stage 1: Srctexts preparation"
+
+    mkdir -p "$(dirname ${other_text})"
+
+    # NOTE(kamo): Give utterance id to each texts.
+    zcat ${WSJ1}/13-32.1/wsj1/doc/lng_modl/lm_train/np_data/{87,88,89}/*.z | \
+	    grep -v "<" | tr "[:lower:]" "[:upper:]" | \
+	    awk '{ printf("wsj1_lng_%07d %s\n",NR,$0) } ' > ${other_text}
+fi
```
Comment (kamo-naoyuki) [inline egs2/dirha_wsj/asr1/local/data.sh]: Could you create nlsyms.txt here?https://github.com/espnet/espnet/blob/2fcade2a8d8f0b910ad7a168510c9403c1bf7362/egs2/wsj/asr1/local/data.sh#L53-L55

File: egs2/dirha_wsj/asr1/path.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
+MAIN_ROOT=$PWD/../../..
+KALDI_ROOT=$MAIN_ROOT/tools/kaldi
+
+[ -f $KALDI_ROOT/tools/env.sh ] && . $KALDI_ROOT/tools/env.sh
```
Comment (kamo-naoyuki) [inline egs2/dirha_wsj/asr1/path.sh]: You don't need to copy this file. Would you change to symbolic link?

File: egs2/dirha_wsj/asr1/run.sh [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+mic=L1C     # Beam_Circular_Array Beam_Linear_Array KA6 L1C
+
+local_data_opts="--mic ${mic}"
+
+stage=1
+stop_stage=11
+nj=16
+decode_asr_model=valid.loss.best.pth 
+
```
Comment (kamo-naoyuki) [inline egs2/dirha_wsj/asr1/run.sh]: Remove these four parameters from run.sh
stage, stop_stage, nj, decode_asr_model because these are not task depended.

**Comment by kamo-naoyuki** [comment]
@yuekaizhang Sorry, I'm not sure whether you'll update anything. Are you finished except for word LM and specaug?
I'd like to merge current PRs even if it's not finished to avoid some conflicts.

File: egs2/dirha_wsj/asr1/conf/train.yaml (hunk 4 above)
Comment (yuekaizhang) [reply egs2/dirha_wsj/asr1/conf/train.yaml]: I deleted the three lines.
max_length:
     - 800  # maxlen_in if input length  > maxlen_in, batchsize is automatically reduced
     - 150

File: egs2/dirha_wsj/asr1/conf/lm.yaml (hunk 5 above)
Comment (yuekaizhang) [reply egs2/dirha_wsj/asr1/conf/lm.yaml]: Changed the name to train_lm.yaml.

File: egs2/dirha_wsj/asr1/local/data.sh (hunk 6 above)
Comment (yuekaizhang) [reply egs2/dirha_wsj/asr1/local/data.sh]: Done.

File: egs2/dirha_wsj/asr1/path.sh (hunk 7 above)
Comment (yuekaizhang) [reply egs2/dirha_wsj/asr1/path.sh]: Done.

File: egs2/dirha_wsj/asr1/run.sh (hunk 8 above)
Comment (yuekaizhang) [reply egs2/dirha_wsj/asr1/run.sh]: Done.

**Comment by yuekaizhang** [comment]
> @yuekaizhang Sorry, I'm not sure whether you'll update anything. Did you finish except for word LM and specaug?
> I'd like to merge current PRs even if it's not finished to avoid some conflicts.

I finished. I only run one type of mics to test the recipe. After word-LM, I could rerun experiments for all kinds of mics.

**Comment by kamo-naoyuki** [comment]
@yuekaizhang Thanks, I'll merge this.
@sw005320 Are you planning to implement word LM for v2 decoding?

**Comment by sw005320** [comment]
> @sw005320 Are you planning to implement word LM for v2 decoding?

I believe we should do it. We may also have some discussions about decoding...

**Comment by kamo-naoyuki** [comment]
@yuekaizhang I forgot to say, we'll make our recipes kaldi-free and kaldi requirement becomes option, so feats_type should be raw (=librosa-fbank) by default. 
It's O.K for the current recipe, but we are also interested in the performance difference of fbank and fbank+pitch for each language task,  so please compare the result at the next time. Thanks!
